src/org/rascalmpl/parser/gtd/SGTDBF.java

```diff
@@ -1325,4 +1469,175 @@ protected T buildResult(AbstractNode result, INodeFlattener<T, S> converter, INo
+	 * Note that this can only be done when we know the actual type of T is IConstructor.
+	 */
+	@SuppressWarnings("unchecked")
+	private T fixErrorNodes(T tree, INodeConstructorFactory<T, S> nodeConstructorFactory) {
+		if (!(tree instanceof IConstructor)) {
+			return tree;
+		}
+
+		return (T) fixErrorNodes((IConstructor) tree, (INodeConstructorFactory<IConstructor, S>) nodeConstructorFactory);
+	}
+
+	private IConstructor fixErrorNodes(IConstructor tree, INodeConstructorFactory<IConstructor, S> nodeConstructorFactory) {
+		IConstructor result;
+		Type type = tree.getConstructorType();
+		if (type == RascalValueFactory.Tree_Appl) {
+			result = fixErrorAppl(tree, nodeConstructorFactory);
+		} else if (type == RascalValueFactory.Tree_Char) {
+			result = tree;
+		} else if (type == RascalValueFactory.Tree_Amb) {
+			result = fixErrorAmb(tree, nodeConstructorFactory);
+		} else if (type == RascalValueFactory.Tree_Cycle) {
+			result = tree;
+		} else {
+			throw new RuntimeException("Unrecognized tree type: " + type);
+		}
+
+		if (result != tree && tree.asWithKeywordParameters().hasParameter(RascalValueFactory.Location)) {
+			IValue loc = tree.asWithKeywordParameters().getParameter(RascalValueFactory.Location);
+			result = result.asWithKeywordParameters().setParameter(RascalValueFactory.Location, loc);
+		}
+
+		return result;
+	}
+
+	private IConstructor fixErrorAppl(IConstructor tree, INodeConstructorFactory<IConstructor, S> nodeConstructorFactory) {
+		IValue prod = tree.get(0);
+		IList childList = (IList) tree.get(1);
+		int childCount = childList.length();
+
+		ArrayList<IConstructor> children = new ArrayList<>(childCount);
+		boolean anyChanges = false;
+		boolean errorTree = false;
+		for (int i=0; i<childCount; i++) {
+			if (i == childCount-1) {
+				// Last child could be a skipped child
+				IConstructor last = (IConstructor) childList.get(childCount-1);
+				if (last.getConstructorType() == RascalValueFactory.Tree_Appl) {
+					IConstructor lastProd = (IConstructor) last.get(0);
+					if (lastProd.getConstructorType() == RascalValueFactory.Production_Skipped) {
+						errorTree = true;
+						children.add(last);
+						break;
+					}
+				}
+			}
+
+			IConstructor child = (IConstructor) childList.get(i);
+			IConstructor resultChild = fixErrorNodes(child, nodeConstructorFactory);
+			children.add(resultChild);
+			if (resultChild != child) {
+				anyChanges = true;
+			}
+		}
+
+		if (errorTree) {
+			return nodeConstructorFactory.createErrorNode(children, prod);
+		} else if (anyChanges) {
+			return nodeConstructorFactory.createSortNode(children, prod);
+		}
+
+		return tree;
+	}
+
+	private IConstructor fixErrorAmb(IConstructor tree, INodeConstructorFactory<IConstructor, S> nodeConstructorFactory) {
+		ISet alternativeSet = (ISet) tree.get(0);
+		ArrayList<IConstructor> alternatives = new ArrayList<>(alternativeSet.size());
+		final AtomicBoolean anyChanges = new AtomicBoolean(false);
+		alternativeSet.forEach(alt -> {
```

Comment:
to avoid the atomic boolean, `ISet` is iterable, so we can also say:

```java
for (IValue alt : alternativeSet) {
....
```

Reply:
Much better!